kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java

```diff
@@ -367,9 +462,35 @@ private SnapshotImpl getSnapshotAtInit(Engine engine)
             getLogSegmentFrom(engine, lastCheckpointOpt);
 
         return logSegmentOpt
-            .map(logSegment -> createSnapshot(
-                logSegment, engine))
-            .orElseThrow(() -> new TableNotFoundException(tablePath.toString()));
+                .map(logSegment -> getCoordinatedCommitsAwareSnapshot(
+                        engine, logSegment, Optional.empty()))
+                .orElseThrow(() -> new TableNotFoundException(tablePath.toString()));
+    }
+
+    /**
+     * This can be optimized by making snapshot hint optimization to work with coordinated commits.
+     * @see <a href="https://github.com/delta-io/delta/issues/3437">issue #3437</a>.
+     */
+    private SnapshotImpl getCoordinatedCommitsAwareSnapshot(
+            Engine engine,
+            LogSegment initialSegmentForNewSnapshot,
+            Optional<Long> versionToLoadOpt) {
+        SnapshotImpl newSnapshot = createSnapshot(initialSegmentForNewSnapshot, engine);
```

Comment:
Can we exit early if the `newSnapshot.version = versionToLoad`?

Reply:
Sure, we can make explicit check. But I think that's the case without CC. So it should exit early with current logic.

Reply:
But if we have CC and are getting a snapshot version that is available on the filesystem we dont need to get anything from CC right?

Reply:
Yes, you are right. Thanks for pointing this out! I'll add early exit.